wordpress-mobile/WordPress-iOS
# Stats Widget: Load stats in the widget after user logs in

## Issue

Closes #17567 

## Description

### Issue

Data on the stats widget is not shown until the user manually visits the stats page.

### Rootcause

There are 2 reasons:
1. App only calls stats widget initialization in `StatsViewController.initStats()`
2. Stat widget itself in `SiteListProvider.getTimeline(...)` doesn't fetch stats while `configuration.site` is nil

### Solution

1. Listen to `WPSigninDidFinishNotification` and initialize stats when the event is received. 
Two important pieces of data are already loaded when this event triggers: 1) `WPStatsHomeWidgetsUserDefaultsSiteIdKey` is set 2) `visibleBlogsForWPComAccounts` are set

2. Update `SiteListProvider` so it would use `defaultSiteID` while `configuration.site` is nil. This allows Stats widget to load required data itself in `SiteListProvider.getTimeline(...)` and not rely on `StatsViewController` being opened for data to be loaded

### Another potential improvement

Create an empty state for the widget while the data is not loaded. For example show the site name and "-" instead of numbers.

## Testing instructions

1st scenario:

1. Open the app and make sure it's logged out
2. Add Widgets
4. Log in
5. Background the app
6. Confirm that Stats load

2nd scenario:

1. Open the app and make sure it's logged out
4. Log in
5. Background the app
7. Add Widgets
8. Confirm that Stats load

3rd scenario:
1. Open the app and make sure it's logged out
2. Log in with the account
3. Add widget
4. Long tap on the widget and click Edit Widget
5. Select a site
6. Come back to the app
7. Logout
8. Log in with the different account
9. Background the app
10. Confirm the Stats load

## Regression Notes

1. Potential unintended areas of impact

Make sure the widget is still updated when visiting Stats view

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

Added `StatsWidgetsStore` to confirm that Stats widget data is loaded or deleted depending on notifications posted. More thorough tests would require a refactoring of `StatsWidgetsStore` to inject WidgetCenter, AccountHelper, HomeWidgetCache, and other dependencies.

## Video

Add widgets before logging in:

https://user-images.githubusercontent.com/4062343/174248260-740ce27d-a9e6-4aa6-a114-60e1b56f048d.mov

Add widgets after logging in:

https://user-images.githubusercontent.com/4062343/173892420-e5ded3ce-09d3-43c3-a60c-c4fdb2b887f2.mov

PR submission checklist:
